### PR TITLE
use nd for accuracy calculation

### LINKED ISSUE #9571: ACCURACY IS USING NUMPY, URGENT FIX
Accuracy is the bottleneck for even a 4000 node neural network. This is a practical job that many people will run and which a GPU ***CAN accelerate***.

But right now the accuracy metric is the bottleneck. That's because it converts everything to numpy, causing a blocking operation at every iteration. Let's fix it stat.

  
```
for label, pred_label in zip(labels, preds):
            if pred_label.shape != label.shape:
                pred_label = ndarray.argmax(pred_label, axis=self.axis)
            pred_label = pred_label.asnumpy().astype('int32')
            label = label.asnumpy().astype('int32')

            check_label_shapes(label, pred_label)

            self.sum_metric += (pred_label.flat == label.flat).sum()
            self.num_inst += len(pred_label.flat)
```

Note: Providing complete information in the most concise form is the best way to get help. This issue template serves as the checklist for essential information to most of the technical issues and bug reports. For non-technical issues and feature requests, feel free to present the information in what you believe is the best form.

For Q & A and discussion, please start a discussion thread at https://discuss.mxnet.io 

## Description
(Brief description of the problem in no more than 2 sentences.)

## Environment info (Required)

```
What to do:
1. Download the diagnosis script from https://raw.githubusercontent.com/apache/incubator-mxnet/master/tools/diagnose.py
2. Run the script using `python diagnose.py` and paste its output here.

```

Package used (Python/R/Scala/Julia):
(I'm using ...)

For Scala user, please provide:
1. Java version: (`java -version`)
2. Maven version: (`mvn -version`)
3. Scala runtime if applicable: (`scala -version`)

For R user, please provide R `sessionInfo()`:

## Build info (Required if built from source)

Compiler (gcc/clang/mingw/visual studio):

MXNet commit hash:
(Paste the output of `git rev-parse HEAD` here.)

Build config:
(Paste the content of config.mk, or the build command.)

## Error Message:
(Paste the complete error message, including stack trace.)

## Minimum reproducible example
(If you are using your own code, please provide a short script that reproduces the error. Otherwise, please provide link to the existing example.)

## Steps to reproduce
(Paste the commands you ran that produced the error.)

1.
2.

## What have you tried to solve it?

1.
2.

### LINKED ISSUE #9582: Misleading calculation of mxnet.metric.Accuracy
## Description
When passes "preds" and "labels" arguments of the same shape to mxnet.metric.Accuracy.update() method, they get silently converted to ints and compared element-wise. This can lead to a weird behavior when "labels" actually are in one-hot encoded format. See the case on StackOverflow: https://stackoverflow.com/questions/45798125/why-mxnet-is-reporting-the-incorrect-validation-accuracy/48471241

I recommend add a warning message if the conversion is happening from floats to int. Removing the conversion is also possible, but it is a breaking change.

## Environment info (Required)

```
----------Python Info----------
Version      : 3.6.2
Compiler     : GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)
Build        : ('default', 'Jul 17 2017 16:44:45')
Arch         : ('64bit', '')
------------Pip Info-----------
Version      : 9.0.1
Directory    : /usr/local/lib/python3.6/site-packages/pip
----------MXNet Info-----------
Version      : 0.11.0
Directory    : /usr/local/lib/python3.6/site-packages/mxnet
Commit Hash   : 53274b4a2b0d73f3fbdb10cfb5f9ed0c8263fda7
----------System Info----------
Platform     : Darwin-16.7.0-x86_64-i386-64bit
system       : Darwin
node         : <removed>
release      : 16.7.0
version      : Darwin Kernel Version 16.7.0: Thu Jan 11 22:59:40 PST 2018; root:xnu-3789.73.8~1/RELEASE_X86_64
----------Hardware Info----------
machine      : x86_64
processor    : i386
b'machdep.cpu.extfeatures: SYSCALL XD 1GBPAGE EM64T LAHF LZCNT PREFETCHW RDTSCP TSCI'
b'machdep.cpu.leaf7_features: SMEP ERMS RDWRFSGS TSC_THREAD_OFFSET BMI1 HLE AVX2 BMI2 INVPCID RTM SMAP RDSEED ADX IPT SGX FPU_CSDS MPX CLFSOPT'
b'machdep.cpu.features: FPU VME DE PSE TSC MSR PAE MCE CX8 APIC SEP MTRR PGE MCA CMOV PAT PSE36 CLFSH DS ACPI MMX FXSR SSE SSE2 SS HTT TM PBE SSE3 PCLMULQDQ DTES64 MON DSCPL VMX SMX EST TM2 SSSE3 FMA CX16 TPR PDCM SSE4.1 SSE4.2 x2APIC MOVBE POPCNT AES PCID XSAVE OSXSAVE SEGLIM64 TSCTMR AVX1.0 RDRAND F16C'
b'machdep.cpu.brand_string: Intel(R) Core(TM) i7-7660U CPU @ 2.50GHz'
----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.0030 sec, LOAD: 0.6859 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.0009 sec, LOAD: 0.1113 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.0015 sec, LOAD: 0.2410 sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0007 sec, LOAD: 0.2651 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0008 sec, LOAD: 0.2126 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0006 sec, LOAD: 0.1419 sec.
```

## Error Message:
There is no visible error message

## Minimum reproducible example
```
import mxnet as mx
predicts = mx.nd.array([[1.29206967e-09,   3.40120096e-05,   2.23299547e-12,   3.98692492e-07,
    1.21151755e-10,   2.59370694e-08,   1.95488334e-02,   1.13474562e-05,
    9.80405331e-01,   3.51648767e-12]])
labels = mx.nd.array([[0, 0, 0, 0, 0, 0, 0, 0, 1, 0]])
acc = mx.metric.Accuracy()
acc.update(preds=predicts, labels=labels)
print(acc.get())
```

which gives us:

`('accuracy', 0.90000000000000002)`

### PR DESCRIPTION
**Pull request by szha**
## Description ##
use ndarray for accuracy calculation.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] use ndarray instead of numpy for `metric.Accuracy` that fixes #9571 
- [x] Improve accuracy doc that fixes #9582 

## Comments ##
- Backward compatible.

### CODE REVIEW

File: python/mxnet/metric.py [hunk 1]
```diff
@@ -380,23 +380,24 @@ def update(self, labels, preds):
...
+            or a vector of likelihoods for all classes.
         """
         check_label_shapes(labels, preds)
 
         for label, pred_label in zip(labels, preds):
             if pred_label.shape != label.shape:
                 pred_label = ndarray.argmax(pred_label, axis=self.axis)
-            pred_label = pred_label.asnumpy().astype('int32')
-            label = label.asnumpy().astype('int32')
+            pred_label = pred_label.astype('int32')
```
Comment (zhreshold) [inline python/mxnet/metric.py]: `int64`is better?

Comment (szha) [reply python/mxnet/metric.py]: This requires larger space can show when the prediction class is large (such as in NLP applications). Should I make it an option?

Comment (zhreshold) [reply python/mxnet/metric.py]: should be good in most cases.

**Comment by piiswrong** [comment]
This is not as simple as changing numpy array to ndarray. See https://github.com/apache/incubator-mxnet/pull/7995
There are some cases where numpy array is faster. Please check performance against all cases.

Also flatten reshapes to 2 dimensions. This could cause problems when output is 1 dimensional. Use reshape((-1,))

**Comment by szha** [comment]
Thanks for the comment. I will switch to use reshape.

Regarding performance, that change was made before the CPU kernel optimization work. Now we should have better performance. Thanks to that, I'm not sure now whether it's worth investing time and code complexity on this now. Also, I'm not entirely convinced that the frontend code should do the job of the backend such as selecting implementation, just for the sake of performance. What do you think?

That said, if there's immediate performance hit due to switching to ND, I'm open to switching back to numpy. Given that it would be infeasible for me to check for all cases, is there any specific observation or reasons from your side that requires attention on its performance?

**Comment by piiswrong** [comment]
Since this is a performance improvement, please verify that it indeed improve performance, at least for common cases.

Please verify this change does not bring back the negative performance impact reported by #7995

**Comment by ptrendx** [comment]
@szha Did you do the performance experiments that @piiswrong asked you about (and what were the results)? With this commit we see 20% perf regression on 8 Voltas in resnet.

**Comment by szha** [comment]
We intend to start tracking the performance of metrics going forward. See 9705 for some numbers. The acc implementation has been updated after this commit. Does the regression still exist on master head?

**Comment by DickJC123** [comment]
We saw the perf regression on a build that includes this PR and also ed823b2 "proper flatten in acc (#9619)."

File: python/mxnet/metric.py [hunk 2]
```diff
@@ -380,23 +380,27 @@ def update(self, labels, preds):
...
+            label = label.astype('int32')
 
             check_label_shapes(label, pred_label)
 
-            self.sum_metric += (pred_label.flat == label.flat).sum()
-            self.num_inst += len(pred_label.flat)
+            if pred_label.context != label.context:
+                pred_label = pred_label.as_in_context(label.context)
+
+            self.sum_metric += (pred_label.flatten() == label.flatten()).sum().asscalar()
```
Comment (ThomasDelteil) [inline python/mxnet/metric.py]: asscalar is equivalent to : asnumpy()[0]

This PR did not solve the problem of the metric being computed in the numpy world

Comment (szha) [reply python/mxnet/metric.py]: Computation happens before asnumpy() happens, so nothing is happening in the numpy world other than passing out a scalar value.

May I ask what your interest is in this PR? Do you have a use case that benefits from using ndarray for metric?

Comment (ThomasDelteil) [reply python/mxnet/metric.py]: Maybe my understanding is flawed but in order to be able to bring back this value on the CPU to self.sum_metric, we will need to do a wait_on_read on the pred_label ?
If we have a loop that has:
```
for batch in data:
   ...
   metric.update()
```
we will wait for the label_pred to be computed before we load the next batch on the GPU.

Whilst if we had `self.sum_metric = nd.zeros(1, ctx=ctx)`
`self.sum_metric += (pred_label.flatten() == label.flatten()).sum()`
then we wouldn't be blocking on this operation before loading the next batch of data and it would just enqueue more operations on the back-end.

I have run several experiments where having the loss computed .asscalar() on every loop is slowing the training by 2-25%.

see:
https://github.com/ilkarman/DeepLearningFrameworks/issues/55
https://github.com/zackchase/mxnet-the-straight-dope/issues/455

Comment (ThomasDelteil) [reply python/mxnet/metric.py]: @szha to precise what I mean, I don't think using numpy is slower in itself, it is just that having a blocking operation in the loop is limiting the use of the parallelization happening in the backend and lead to GPU starvation.

Comment (sxjscience) [reply python/mxnet/metric.py]: However, completely using the non-blocking logic will cause some other problems. To be more specific, the allocated NDArrays cannot be reused and will finally cause an OOM. We should use `asscalar()` to avoid this.

Comment (ThomasDelteil) [reply python/mxnet/metric.py]: Thanks for the precision @sxjscience , bit why is it so? Once processed in the computation graph aren't they garbage collected?

Comment (sxjscience) [reply python/mxnet/metric.py]: I think it's because the OPs are pushed in a much faster speed than the real computation. The graph will keep expanding and the allocation OPs will be executed to allocate new space (Even before the actual computation is performed). We have to call a blocking operator at some point to make sure that the current calculation in the graph has been completed. CC @piiswrong for this.

Comment (ThomasDelteil) [reply python/mxnet/metric.py]: Thanks yes that's my understanding. However I think it should be left to the user to decide when to block, since it depends highly on their GPU and model size (like every 100 batches or every epoch).  Also is there a reason why the accuracy is stored on the CPU rather than on specific context? My measures showed great improvements when storing the accuracy on GPU. Maybe if you don't mind we can continue the discussion there: https://github.com/apache/incubator-mxnet/issues/9571

**Comment by sxjscience** [comment]
Sure. Let’s move the discussion there.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
